CRM/Core/BAO/MessageTemplate.php

```diff
@@ -583,21 +581,16 @@ protected static function loadTemplate(string $workflowName, bool $isTest, int $
    * @return array
    */
   public static function renderMessageTemplate(array $mailContent, bool $disableSmarty, $contactID, array $smartyAssigns): array {
```

Comment:
@totten @eileenmcnaughton I asked for `$contactID` to be an array when this function was added because that is what it will need to be even if we don't yet support many things (but note if you pass [`activityId` => 123, `contactId` => 234] that will already work for activity tokens via tokenProcessor).  There is an unresolved question of how we should support multiple IDs (eg. multiple activities, multiple contacts).

Reply:
Yeah - we can change this signature when we are ready/ when we have the test cover to be comfortable extending it. I think we have been pretty clear it shouldn't be called from outside of core

Reply:
Right, multiple IDs are interesting. I would distinguish between a few cases, though (easiest to hardest):

1. Handling one contact
    * Ex: `{contact.*}` <=> `$tokenContext['contactId']=123`
    * Before this patch, `renderMessageTemplate(int $contactID)`  allowed one contact.
2. Handling one of each entity (one contact, one activity, one contribution, etc)
    * Ex: `{activity.*}` <=> `$tokenContext['activityId']=123`
    * With this patch, `render(array $tokenContext)` allows one contact, one activity, one contribution, etc.
3. Handling multiple instances of the same entity - each in different names/roles/relations
    * Ex:`{assignee.*}` <=> `$tokenContext['assigneeContactId']=456`, `{last_contribution.status}` <=> `$tokenContext['lastContributionId']=456`
    * With or without this patch, the main constraint is the `CRM_*_Tokens` listeners. Those could be updated incrementally/independently. If `CRM_*_Tokens` can accept distinct names/roles/relations, then the dataflow from `render(array $tokenContext)` => `CRM_*_Tokens` will work fine.
4. Handling lists - ie multiple instances having the same type+role
    * Ex: `$tokenContext['familyMemberIds']=[123,456,789]` <=> *no clear token notation*, `$tokenContext['contributionHistory']=[123,456,789]` <=> *no clear token notation*, 
    * With or without this patch, the Civi token-language is the constraint. This is a harder topic, and IMHO the techniques will be very different.